scipy/interpolate/_ndbspline.py

```diff
@@ -0,0 +1,210 @@
+import operator
+import numpy as np
+
+from math import prod
+
+from . import _bspl
+
+__all__ = ["NdBSpline"]
+
+
+def _get_dtype(dtype):
+    """Return np.complex128 for complex dtypes, np.float64 otherwise."""
+    if np.issubdtype(dtype, np.complexfloating):
+        return np.complex_
+    else:
+        return np.float_
+
+
+class NdBSpline:
+    """Tensor product spline object.
+
+    The value at point ``xp = (x1, x2, ..., xN)`` is evaluated as a linear
+    combination of products of one-dimensional b-splines in each of the ``N``
+    dimensions::
+
+       c[i1, i2, ..., iN] * B(x1; i1, t1) * B(x2; i2, t2) * ... * B(xN; iN, tN)
+
+
+    Here ``B(x; i, t)`` is the ``i``-th b-spline defined by the knot vector
+    ``t`` evaluated at ``x``.
+
+    Parameters
+    ----------
+    c : ndarray, shape (n1, n2, ..., nN, ...)
+        b-spline coefficients
+    t : tuple of 1D ndarrays
+        knot vectors in directions 1, 2, ... N,
+        ``len(t[i]) == n[i] + k + 1``
+    k : int or length-d tuple of integers
+        spline degrees.
+        A single integer is interpreted as having this degree for
+        all dimensions.
+    extrapolate : bool, optional
+        Whether to exrapolate out-of-bounds inputs, or return `nan`.
+        Default is to extrapolate.
+
+    Attributes
+    ----------
+    t : tuple of ndarrays
+        Knots vectors.
+    c : ndarray
+        Coefficients of the tensor-produce spline.
+    k : tuple of integers
+        Degrees for each dimension.
+    extrapolate : bool
+        Whether to extrapolate or return nans for out-of-bounds inputs.
+
+    Methods
+    -------
+    __call__
+
+    See Also
+    --------
+    BSpline : a one-dimensional B-spline object
+    NdPPoly : an N-dimensional piecewise tensor product polynomal
+
+    """
+    def __init__(self, t, c, k, extrapolate=None):
+        ndim = len(t)
+
+        try:
+            len(k)
+        except TypeError:
+            # make k a tuple
+            k = (k,)*ndim
+
+        if len(k) != ndim:
+            raise ValueError(f"len(t) = {ndim} != {len(k)} = len(k)")
```

Comment:
FYI: this can be simplified a little bit by using `=` in the expression. See [this](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging) entry in the Python 3.8 release notes. So now the expression could be `f"{len(t) = } != {len(k) = }"`. 

I'm not sure if it's any neater but it could be something you didn't know of.

Reply:
How, I did not know about it. Thank you! Have used it in the last update.